File: Sudoku.rb
UNIT  = 3
VIEW  = 4
SIZE  = UNIT * UNIT
COORD = SIZE * SIZE * SIZE
BLOCK = VIEW * SIZE * SIZE

module View
  GRD = 0
  ROW = 1
  COL = 2
  BOX = 3
end

module State
  OPEN   = 0
  FIXED  = 1
  BANNED = 2
end

DONE = SIZE + 1

DIGITS  = "123456789"

def coord(i, j, k)
  return (i * SIZE + j) * SIZE + k
end

def block(v, p, q)
  return (v * SIZE + p) * SIZE + q
end

PARENTS = Array.new(COORD)
SIZE.times do |i|
  SIZE.times do |j|
    p = i / UNIT * UNIT + j / UNIT
    SIZE.times do |k|
      PARENTS[coord(i, j, k)] = [
        block(View::GRD, i, j),
        block(View::ROW, i, k),
        block(View::COL, j, k),
        block(View::BOX, p, k),
      ]
    end
  end
end

CHILDREN = Array.new(BLOCK)
BLOCK.times do |b|
  CHILDREN[b] = []
end
COORD.times do |c|
  PARENTS[c].each do |b|
    CHILDREN[b].push(c)
  end
end

class NoSolutionError < StandardError
end

class Sudoku

  attr_accessor :state, :count, :queue

  def initialize(that = nil)
    if that == nil
      @state = Array.new(COORD, State::OPEN)
      @count = Array.new(BLOCK, SIZE)
    else
      @state = that.state.dup
      @count = that.count.dup
    end
    @queue = Array.new
  end

  def enqueue(c)
    @queue.push(c)
    return self
  end

  def read(input)
    SIZE.times do |i|
      SIZE.times do |j|
        if i * SIZE + j < input.length
          k = DIGITS.index(input[i * SIZE + j])
          enqueue(coord(i, j, k)) unless k.nil?
        end
      end
    end
    return self
  end

  def to_s
    output = Array.new(SIZE * SIZE, ".")
    SIZE.times do |i|
      SIZE.times do |j|
        SIZE.times do |k|
          if @state[coord(i, j, k)] == State::FIXED
            output[i * SIZE + j] = DIGITS[k]
          end
        end
      end
    end
    output.join
  end

  def search

    until @queue.empty?
      c = @queue.shift
      case @state[c]
      when State::OPEN
        fix(c)
      when State::FIXED
        nil
      when State::BANNED
        raise NoSolutionError.new
      end
    end

    m = @count.min
    return self if m == DONE

    b = @count.index(m)
    CHILDREN[b].each do |c|
      if @state[c] == State::OPEN
        begin
          return Sudoku.new(self).enqueue(c).search
        rescue NoSolutionError
        end
      end
    end

    raise NoSolutionError.new
  end

  def fix(c)
    @state[c] = State::FIXED
    PARENTS[c].each { |b| mark(b) }
  end

  def mark(b)
    @count[b] = DONE
    CHILDREN[b].each { |c| ban(c) if @state[c] == State::OPEN }
  end

  def ban(c)
    @state[c] = State::BANNED
    PARENTS[c].each { |b| countdown(b) if @count[b] != DONE }
  end

  def countdown(b)
    @count[b] -= 1
    case @count[b]
    when 0
      raise NoSolutionError.new
    when 1
      CHILDREN[b].each { |c| enqueue(c) if @state[c] == State::OPEN }
    end
  end
end

def solve(input)
  begin
    return Sudoku.new.read(input).search.to_s
  rescue NoSolutionError
    return "NO SOLUTION"
  end
end

loop do
  input = $stdin.gets
  break if input.nil?
  puts solve(input.chomp)
end

File: Sudoku.jl
M = 3
N = M * M

N_COORD = N * N * N
N_BLOCK = 4 * N * N

coord(i, j, k) = ((i - 1) * N + j - 1) * N + k

block(i, j, k) = begin
    p = div(i - 1, M) * M + div(j - 1, M) + 1
    [ (0 * N + i - 1) * N + j
    , (1 * N + i - 1) * N + k
    , (2 * N + j - 1) * N + k
    , (3 * N + p - 1) * N + k
    ]
end

parents = zeros(Int, N_COORD, 4)
for i in 1:N, j in 1:N, k in 1:N
    parents[coord(i, j, k), :] = block(i, j, k)
end

children = zeros(Int, N_BLOCK, N)
for b in 1:N_BLOCK
    children[b, :] = find(any(parents .== b, 2))
end

DIGITS = "123456789"

OPEN   = 0
FIXED  = 1
BANNED = 2

CLOSED = N + 1

Coord = Int
Block = Int
State = Int
Count = Int

struct Sudoku
    state::Array{State, 1}
    count::Array{Count, 1}
    queue::Array{Coord, 1}
end
Sudoku() = begin
    state = fill(OPEN, N_COORD)
    count = fill(N, N_BLOCK)
    queue = Coord[]
    Sudoku(state, count, queue)
end
Sudoku(s::Sudoku) = begin
    Sudoku(copy(s.state), copy(s.count), copy(s.queue))
end
Sudoku(str::String) = begin
    s = Sudoku()
    for i in 1:N, j in 1:N
        ij = (i - 1) * N + j
        if ij <= length(str)
            k = search(DIGITS, str[ij])
            if k != 0
                enqueue!(s, coord(i, j, k))
            end
        end
    end
    s
end

struct NoSolutionException <: Exception end

show(s::Sudoku) = begin
    res = fill('.', N, N)
    for i in 1:N, j in 1:N, k in 1:N
        if s.state[coord(i, j, k)] == FIXED
            res[j, i] = DIGITS[k]
        end
    end
    join(res)
end

enqueue!(s::Sudoku, c::Coord) = begin
    push!(s.queue, c)
    s
end

solve!(s::Sudoku) = begin

    while !isempty(s.queue)
        c = shift!(s.queue)
        fix!(s, c)
    end

    m, b = findmin(s.count)
    if m == CLOSED
        return s
    end

    for c in children[b, :]
        if s.state[c] == OPEN
            try
                return solve!(enqueue!(Sudoku(s), c))
            catch ME
                if !(ME isa NoSolutionException)
                    throw(ME)
                end
            end
        end
    end

    throw(NoSolutionException())
end

fix!(s::Sudoku, c::Coord) = begin
    if s.state[c] == FIXED
        return
    elseif s.state[c] == BANNED
        throw(NoSolutionException())
    end
    s.state[c] = FIXED
    for b in parents[c, :]
        mark!(s, b)
    end
end

mark!(s::Sudoku, b::Block) = begin
    if s.count[b] == CLOSED
        throw(NoSolutionException())
    end
    s.count[b] = CLOSED
    for c in children[b, :]
        ban!(s, c)
    end
end

ban!(s::Sudoku, c::Coord) = begin
    if s.state[c] != OPEN
        return
    end
    s.state[c] = BANNED
    for b in parents[c, :]
        countdown!(s, b)
    end
end

countdown!(s::Sudoku, b::Block) = begin
    if s.count[b] == CLOSED
        return
    end
    s.count[b] -= 1
    if s.count[b] == 0
        throw(NoSolutionException())
    elseif s.count[b] == 1
        for c in children[b, :]
            if s.state[c] == OPEN
                enqueue!(s, c)
            end
        end
    end
end

for str in eachline()
    try
        println(show(solve!(Sudoku(str))))
    catch ME
        if ME isa NoSolutionException
            println("NO SOLUTION")
        else
            throw(ME)
        end
    end
end
